Document alerting (#2570)
giantswarm/docs: src/content/tutorials/observability/alerting/create-rules/_index.md

Here is how you would define your alerting rule:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  labels:
    # This lets Alloy know to which tenant this alert belongs to
    observability.giantswarm.io/tenant: my-team
  name: cluster-resource-usage-too-high
  namespace: my-namespace
spec:
  groups:
  - name: reliability
    rules:
      - alert: ComponentDown
        annotations:
          # See https://grafana.com/docs/grafana/latest/alerting/alerting-rules/templates/#template-annotations for useful annotations used by Grafana.
          __dashboardUid__: my-dashboard-uid
          summary: 'Component is done.'
          description: 'Component {{ $labels.service }} is down.'
          runbook_url: https://my-pretty-runbook
        expr: count(up{job=~"component/.*"} == 0) by (cluster_id) > 0
        # Alert if the expression returns results for more than 5 minutes
        for: 5m
        labels:
          # This is a paging alert that need to wake my team up
          severity: page
```

> ## Differenciate between metric vs Log-based rules

Author: ```suggestion
## Differentiate between metric vs Log-based rules
```

Metric-based rules are evaluated by the Mimir ruler and are written using PromQL while Log-based rules are evaluated by the Loki ruler and are written using LogQL.

To be able to differentiate if your alert is a log-based alert, you should label you alert with those labels:

```yaml
# This label is marked as deprecated but still needs to be configured.
application.giantswarm.io/prometheus-rule-kind: loki
observability.giantswarm.io/rule-type: logs
```

## Cluster vs installation scoping of rules

To avoid conflicts when loading rules from multiple clusters in the same tenant (for instance, if your teams define a PrometheusRule in their application template deployed in a dev, staging and prod clusters), the Observability Platform support some level of scoping.

Any rule deployed in workload clusters is by default cluster-scoped (e.g only target metrics belonging to this cluster) while a rule deployed in management clusters is scoped to the installation (e.g. target all metrics of all clusters).

In practice, this means that:

> - If you deploy a rule with the expression `up{job="good"} > 0` in the workload cluster named `alpha1`, then the loaded rules will have the expression `up{cluster_id="alpha1", job="good"} > 0`.

Author: ```suggestion

- If you deploy a rule with the expression `up{job="good"} > 0` in the workload cluster named `alpha1`, then the loaded rules will have the expression `up{cluster_id="alpha1", job="good"} > 0`.
```

- If you deploy a rule with the expression `up{job="good"} > 0` in the _management cluster_ named `alpha1`, then the loaded rules will have the expression `up{job="good"} > 0`.

### Limitations

- Cluster-scope is only available to metric-based alerts today due to upstream limitations

> - Scoping can only be used when teams deploy the same rule __once__ accross multiple clusters. If you need to deploy it per app in different namespaces, you will need to manage the conflict yourself.

Author: ```suggestion
- Scoping can only be used when teams deploy the same rule **once** accross multiple clusters. If you need to deploy it per app in different namespaces, you will need to manage the conflict yourself.
```

---
linkTitle: Creating alerting rules

> title: How to create alerts and recording rules

Reviewer: ```suggestion
title: How to create alerting and recording rules
```

> description: Guide explaining how to create alerts and recording rules in the Observability Platform.

Reviewer: ```suggestion
description: Guide explaining how to create alerting and recording rules in the Observability Platform.
```

menu:
  principal:
    identifier: tutorials-observability-alerting-alerts
    parent: tutorials-observability-alerting
weight: 40
last_review_date: 2025-04-24
user_questions:

>   - How to create alerts?

Reviewer: ```suggestion
  - How to create alerting rules?
```

  - How to create recording rules?

>   - What is log-based alerting?

Reviewer: ```suggestion
  - What is log-based alerting?
  - How to get alerts?
```

owner:
  - https://github.com/orgs/giantswarm/teams/team-atlas
---

The Giant Swarm Observability Platform provides an [alerting pipeline]({{< relref "/overview/observability/alerting/" >}}) that [you can configure per tenant]({{< relref "/tutorials/observability/alerting/configure-alertmanager" >}}) as well as allow you to create your own alerting and recording rules per tenant.

As Giant Swarm embraces GitOps, alerting and recording rules needs to be defined via [Prometheus Operator](https://prometheus-operator.dev/) PrometheusRules. Those rules can be defined in both your management cluster and workload clusters alike.

**Warning:** As our multi-tenancy aligns tenants across our platform on Grafana Organizations please make sure that the `observability.giantswarm.io/tenant` label defined on your rules references an existing tenant defined in a Grafana Organization. Any PrometheusRules that references an non-existing tenant will be ignored. Learn more about our multi-tenancy in [Multi-tenancy in the observability platform]({{< relref "/tutorials/observability/multi-tenancy/" >}})

> ## How to define a recording rule

Reviewer: Isn't like an alerting rule more often used than a recording rule? Shouldn't we be doing alerting rule, then recording rule? idk

Author: Maybe, I ordered them that way because recording rules are easier to understand

Author: Order is switched

          runbook_url: https://my-pretty-runbook
        expr: count(up{job=~"component/.*"} == 0) by (cluster_id) > 0
        # Alert if the expression returns results for more than 5 minutes
        for: 5m
        labels:
          # This is a paging alert that need to wake my team up
          severity: page
```

## Differentiate between metric vs Log-based rules

Metric-based rules are evaluated by the Mimir ruler and are written using PromQL while Log-based rules are evaluated by the Loki ruler and are written using LogQL.

To be able to differentiate if your alert is a log-based alert, you should label you alert with those labels:

```yaml
# This label is marked as deprecated but still needs to be configured.
application.giantswarm.io/prometheus-rule-kind: loki
observability.giantswarm.io/rule-type: logs
```

## Cluster vs installation scoping of rules

To avoid conflicts when loading rules from multiple clusters in the same tenant (for instance, if your teams define a PrometheusRule in their application template deployed in a dev, staging and prod clusters), the Observability Platform support some level of scoping.

> Any rule deployed in workload clusters is by default cluster-scoped (e.g only target metrics belonging to this cluster) while a rule deployed in management clusters is scoped to the installation (e.g. target all metrics of all clusters).

Author: ```suggestion
Any rule deployed in workload clusters is by default cluster-scoped (only targets metrics coming from this cluster) while a rule deployed in management clusters is scoped to the installation (targets all metrics of all clusters).
```

---
linkTitle: Creating alerting rules
title: How to create alerting and recording rules
description: Guide explaining how to create alerting and recording rules in the Observability Platform.
menu:
  principal:
    identifier: tutorials-observability-alerting-alerts
    parent: tutorials-observability-alerting
weight: 40
last_review_date: 2025-04-24
user_questions:
  - How to create alerting rules?
  - How to create recording rules?
  - What is log-based alerting?
  - How to get alerts?
owner:
  - https://github.com/orgs/giantswarm/teams/team-atlas
---

The Giant Swarm Observability Platform provides an [alerting pipeline]({{< relref "/overview/observability/alerting/" >}}) that [you can configure per tenant]({{< relref "/tutorials/observability/alerting/configure-alertmanager" >}}) as well as allow you to create your own alerting and recording rules per tenant.

As Giant Swarm embraces GitOps, alerting and recording rules needs to be defined via [Prometheus Operator](https://prometheus-operator.dev/) PrometheusRules. Those rules can be defined in both your management cluster and workload clusters alike.

> **Warning:** As our multi-tenancy aligns tenants across our platform on Grafana Organizations please make sure that the `observability.giantswarm.io/tenant` label defined on your rules references an existing tenant defined in a Grafana Organization. Any PrometheusRules that references an non-existing tenant will be ignored. Learn more about our multi-tenancy in [Multi-tenancy in the observability platform]({{< relref "/tutorials/observability/multi-tenancy/" >}})

Reviewer 2: ```suggestion
**Warning**: As our multi-tenancy aligns tenants across our platform on Grafana Organizations please make sure that the `observability.giantswarm.io/tenant` label defined on your rules references an existing tenant defined in a Grafana Organization. Any PrometheusRules that references an non-existing tenant will be ignored. Learn more about our multi-tenancy in [Multi-tenancy in the observability platform]({{< relref "/tutorials/observability/multi-tenancy/" >}})
```

        annotations:
          # See https://grafana.com/docs/grafana/latest/alerting/alerting-rules/templates/#template-annotations for useful annotations used by Grafana.
          __dashboardUid__: my-dashboard-uid
          summary: 'Component is down.'
          description: 'Component {{ $labels.service }} is down.'
          runbook_url: https://my-pretty-runbook
        expr: count(up{job=~"component/.*"} == 0) by (cluster_id) > 0
        # Alert if the expression returns results for more than 5 minutes
        for: 5m
        labels:
          # This is a paging alert that need to wake my team up
          severity: page
```

## Differentiate between metric vs Log-based rules

Metric-based rules are evaluated by the Mimir ruler and are written using PromQL while Log-based rules are evaluated by the Loki ruler and are written using LogQL.

To be able to differentiate if your alert is a log-based alert, you should label you alert with those labels:

```yaml
# This label is marked as deprecated but still needs to be configured.
application.giantswarm.io/prometheus-rule-kind: loki
observability.giantswarm.io/rule-type: logs
```

## Cluster vs installation scoping of rules

> To avoid conflicts when loading rules from multiple clusters in the same tenant (for instance, if your teams define a PrometheusRule in their application template deployed in a dev, staging and prod clusters), the Observability Platform support some level of scoping.
> 
> Any rule deployed in workload clusters is by default cluster-scoped (only targets metrics coming from this cluster) while a rule deployed in management clusters is scoped to the installation (targets all metrics of all clusters).

Reviewer 2: ```suggestion
The Observability Platform supports some level of scoping to avoid conflicts when loading rules from multiple clusters in the same tenant (for instance, if your teams define a PrometheusRule in their application template deployed in dev, staging, and prod clusters).

Any rule deployed in workload clusters is, by default, cluster-scoped (it only targets metrics coming from this cluster), while a rule deployed in management clusters is scoped to the installation (it targets all metrics of all clusters).
```
